### PR TITLE
Adds resource revisions to applications.

### PR DESCRIPTION
**Pull request by alesstimec**
## Description

Allows users to specify exact resource revisions when deploying an application.

Fixes: #322

## Type of change

- Change existing resource

## Environment

- Juju controller version: 

- Terraform version: 

## QA steps

Manual QA steps should be done to test this PR.

```tf
provider juju {}
...
```

### CODE REVIEW

**Review by hmlanigan — COMMENTED** [review]
Thank you for the bug fix. A few initial thoughts.

File: internal/provider/resource_application_test.go [hunk 1]
```diff
@@ -203,6 +203,74 @@ func TestAcc_CharmUpdates(t *testing.T) {
...
+		t.Skip(t.Name() + " only runs with LXD")
+	}
+	modelName := acctest.RandomWithPrefix("tf-test-resource-revision-updates")
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:                 func() { testAccPreCheck(t) },
+		ProtoV6ProviderFactories: frameworkProviderFactories,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccResourceApplicationUpdatesResource(modelName, "candid", "latest/candidate", "candid", "3"),
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply internal/provider/resource_application_test.go]: perfect, thank you.. spent too much time looking for a charm i could use in tests..

File: internal/provider/resource_application.go [hunk 2]
```diff
@@ -174,6 +176,11 @@ func (r *applicationResource) Schema(_ context.Context, _ resource.SchemaRequest
 					stringplanmodifier.UseStateForUnknown(),
 				},
 			},
+			ResourceKey: schema.MapAttribute{
```
[the remark this answers is not in the document]
Comment (alesstimec) [reply internal/provider/resource_application.go:179]: added issue #428

File: docs/resources/application.md [hunk 3]
```diff
@@ -66,6 +66,17 @@ resource "juju_application" "placement_example" {
...
 - `placement` (String) Specify the target location for the application's units
+- `resources` (Map of Number) Charm resource revisions. Must evaluate to an integer.
+
+	There are a few scenarios that need to be considered:
+	* If the plan does not specify resource revision and resources are added to the plan,
+	resources with specified revisions will be attached to the application (equivalent
+	to juju attach-resource).
+	* If the plan does specify resource revisions and:
+		* If the charm revision or channel is updated, then resources get updated to the 
+		  latest revision.
```
Comment (hmlanigan) [inline docs/resources/application.md:77]:
```suggestion
		* If the charm revision or channel is updated, then resources do NOT get updated to the 
		  latest revision.
```

File: docs/resources/application.md [hunk 4]
```diff
@@ -66,6 +66,17 @@ resource "juju_application" "placement_example" {
...
+		* If the charm revision or channel are not updated, then no changes will take 
+		  place (juju does not have an "un-attach" command for resources).
```
Comment (hmlanigan) [inline docs/resources/application.md:79]:
```suggestion
	* If you remove resource revisions from a plan, they will not be removed from the application. (juju does not have an "un-attach" command for resources).
```

File: internal/provider/resource_application_test.go [hunk 5]
```diff
@@ -203,6 +203,74 @@ func TestAcc_CharmUpdates(t *testing.T) {
...
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("juju_application.juju-qa-test", "resources.foo-file", "3"),
+				),
+			},
+			{
+				// change back to 4
+				Config: testAccResourceApplicationWithRevisionAndConfig(modelName, "juju-qa-test", 21, "", "foo-file", 4),
+				Check: resource.ComposeTestCheckFunc(
+					resource.TestCheckResourceAttr("juju_application.juju-qa-test", "resources.foo-file", "4"),
+				),
```
Comment (hmlanigan) [inline internal/provider/resource_application_test.go:234]: There is another scenario to cover to replace this. Removing the resource revision of the plan, updating the charm to a new channel, while removing the revision.

**Review by hmlanigan — CHANGES_REQUESTED** [review]
Over all looks good. Thank you.

I did some QA on my own, however the QA steps do need to filled out in the PR description.

Just a few things to fix.

File: internal/provider/resource_application_test.go (hunk 5 above)
Comment (alesstimec) [reply internal/provider/resource_application_test.go:234]: added two tests:

1. starts with no resource revision in the plan, then adds it
2. starts with resource revision in the plan, then removes it

but.. the issue is that i cannot assert what juju is given in the update command, just that the state does not contain a resource revision definition..

Comment (hmlanigan) [reply internal/provider/resource_application_test.go:234]: we're working to add more unit tests rather than only the acceptance tests.

Comment (hmlanigan) [reply internal/provider/resource_application_test.go:234]: The Acceptance tests also run Read after each step and compare.

**Comment by hmlanigan** [comment]
resolved conflicts after other PRs landed.

**Review by hmlanigan — APPROVED** [review]
LGTM, Thank you for the contribution!

**Pull request merged**